syntax = "proto3";

package thoughts;

// ---------------- General ----------------

message Empty {}

message DataRequest {
  int32 user_id = 1;
  string token = 2;
  int32 page = 3;
  int32 limit = 4;
}

message Identifiers {
  repeated int32 ids = 1;
}

// ------------ Authentication ------------

message Credentials {
  string username = 1;
  string password = 2;
  string user_agent = 3;
  string email = 4;
}

message AuthRequest {
  string token = 1;
}

message AuthStatus {
  string token_type = 1;
  string access_token = 2;
  string refresh_token = 3;
  int32 user_id = 4;
  int32 session_id = 5;
}

// ---------------- Users ----------------

message UserRequest {
  int32 user_id = 1;
  string username = 2;
  string token = 3;
}

message Users {
  repeated User users = 1;
}

message UserUpdates {
  string username = 1;
  string email = 2;
  string name = 3;
  string password = 4;
  string bio = 5;
  string old_password = 6;
  string avatar = 7;
  string token = 8;
}

message User {
  int32 id = 1;
  string username = 2;
  string email = 3;
  string name = 4;
  string bio = 5;
  string avatar = 6;
  int32 posts = 7;
  int32 likes = 8;
  int32 following = 9;
  int32 followers = 10;
  string date_created = 11;
}

// ---------------- Sessions -----------------

message Session {
  int32 id = 1;
  string name = 2;
  string user_agent = 3;
  int32 user_id = 4;
  string date_created = 5;
}

message Sessions {
  repeated Session sessions = 1;
}

message SessionRequest {
  int32 session_id = 1;
  string token = 2;
}

// ---------------- Posts ----------------

/**
 * Posts
 */

message Post {
  int32 id = 1;
  string content = 2;
  int32 user_id = 3;
  int32 likes = 4;
  bool liked = 5;
  int32 retweets = 6;
  bool retweeted = 7;
  string date_created = 8;
}

message PostUpdates {
  int32 post_id = 1;
  string content = 2;
  string token = 3;
}

message Posts {
  repeated Post posts = 1;
}

message PostRequest {
  int32 post_id = 1;
  string token = 2;
}

// ---------------- Services ----------------

service UserService {
  rpc CreateUser(UserUpdates) returns (User);
  rpc GetUser(UserRequest) returns (User);
  rpc UpdateUser(UserUpdates) returns (Empty);
  rpc UpdatePassword(UserUpdates) returns (Empty);
  rpc DeleteUser(UserRequest) returns (Empty);
}

service FollowService {
  rpc GetFollowing(DataRequest) returns (Users);
  rpc GetFollowingIds(UserRequest) returns (Identifiers);
  rpc GetFollowers(DataRequest) returns (Users);
  rpc GetFollowersIds(UserRequest) returns (Identifiers);
  rpc FollowUser(UserRequest) returns (Empty);
  rpc UnfollowUser(UserRequest) returns (Empty);
}

service AuthService {
  rpc Login(Credentials) returns (AuthStatus);
  rpc Refresh(AuthRequest) returns (AuthStatus);
  rpc Validate(AuthRequest) returns (AuthStatus);
  rpc ValidatePassword(Credentials) returns (Empty);
}

service SessionService {
  rpc GetSessions(AuthRequest) returns (Sessions);
  rpc DeleteSession(SessionRequest) returns (Empty);
}

service PostService {
  rpc CreatePost(PostUpdates) returns (Post);
  rpc GetPost(PostRequest) returns (Post);
  rpc GetFeed(DataRequest) returns (Posts);
  rpc GetPosts(DataRequest) returns (Posts);
  rpc GetLikedPosts(DataRequest) returns (Posts);
  rpc DeletePost(PostRequest) returns (Empty);
}

service ActionService {
  rpc LikePost(PostRequest) returns (Empty);
  rpc UnlikePost(PostRequest) returns (Empty);
  rpc RetweetPost(PostRequest) returns (Empty);
  rpc RemoveRetweet(PostRequest) returns (Empty);
}
